To be meged for MTV 2.8.0: MTV-2029 Migrating VMs with shared disks (#643)
kubev2v/forklift-documentation: documentation/modules/mtv-shared-disks.adoc

// Module included in the following assemblies:
//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_mod-docs-content-type: PROCEDURE
[id="mtv-shared-disks_{context}"]
= Migrating virtual machines with shared disks

You can migrate VMware virtual machines (VMs) with shared disks using the {project-first}. This functionality is available for cold migrations only. 

> Shared disks are disks that are attached to multiple VMs and that use the multi-writer option. As a result of these characteristics, shared disks are difficult to migrate. 

Reviewer: Use simple language. Consider using 'many' rather than 'multiple'.Vale[RedHat.SimpleWords](https://redhat-documentation.github.io/vale-at-red-hat/docs/main/reference-guide/simplewords/)

> In certain situations, applications in VMs require shared disks. Databases and clustered file-systems are the primary use cases for shared disks.

Reviewer: ```suggestion
In certain situations, applications in VMs require shared disks. Databases and clustered file systems are the primary use cases for shared disks.
```

> {project-short} version 2.7.11 or later includes a parameter named `migrateSharedDisks` in `Plan` CRs that instructs {project-short} to either migrate shared disks or to skip them during migration as follows:

Reviewer: ```suggestion
{project-short} version 2.7.11 or later includes a parameter named `migrateSharedDisks` in `Plan` CRs that instructs {project-short} to either migrate shared disks or to skip them during migration, as follows:
```

Reviewer: CRs  = custom resources (CRs)

* If set to `true`, {project-short} uses the regular cold migration flow using `virt-v2v` and labeling the shared PVC. This is the default setting. 

> * If set to `false`, {project-short} uses the KubeVirt CDI for disk transfer, and skips the shared disks. After the disk transfer, {project-short} automatically attempts to locate the already shared PVCs and the already migrated shared disks and attach them to the VMs. 

Reviewer: ```suggestion
* If set to `false`, {project-short} uses the KubeVirt CDI for disk transfer and skips the shared disks. After the disk transfer, {project-short} automatically attempts to locate the already shared PVCs and the already migrated shared disks and attach them to the VMs. 
```

> * If set to `false`, {project-short} uses the KubeVirt CDI for disk transfer, and skips the shared disks. After the disk transfer, {project-short} automatically attempts to locate the already shared PVCs and the already migrated shared disks and attach them to the VMs. 

Reviewer: PVCs - persistent volume claims (PVCs)

By default, `migrateSharedDisks` is set to `true`.

To successfully migrate VMs with shared disks, create two `Plan` CRs as follows: 

* In the first, set `migrateSharedDisks` to `true`. {project-short} migrates the following:

** All shared disks.

> ** For each shared disk, one of the VMs that is attached to it. If possible, choose VMs so that the plan does not contain any shared disks that are connected to more than one VM in the plan. See the figures below for further guidance.

Reviewer: Please do not use `below`. Use `following`

> ** For each shared disk, one of the VMs that is attached to it. If possible, choose VMs so that that plan does not contain any shared disks that are connected to more than one VM in the plan. See the figures below for further guidance.

Reviewer 2: Typo in `choose VMs so that that` -> `choose VMs so that`?

** All unshared disks attached to the VMs you choose for this plan. 

> * In the second, set `migrateSharedDisks` to 'false`. {project-short} migrates the following:

Reviewer 2: 'false\` -> \`false\`

** All other VMs
** The unshared disks of the VMs in the second `Plan` CR

To illustrate the division of VMs and disks between `Plan` CRs, consider the two figures that follow: 

> In the first figure, the first `Plan` CR migrates VMs 1 and 3, shared disks 1, 2, and 3, and the non-shared disks of VMs 1 and 3. VMs 1 and 3 were chosen because together they connect to all the shared disks. In practice, {project-short} first migrates VM 1 and all disks attached to it, then does the same with VM 3.Choosing VM 1 and VM 3 and only those two ensures that all the shared disks are migrated, but only once.  

Reviewer 2: This is missing the reference to the figure

Author: This figure will be replaced to fit documentation requirements for figures.

> In the first figure, the first `Plan` CR migrates VMs 2 and 4, shared disks 1, 2, and 3, and the non-shared disks of VMs 2 and 4. VMs 2 and 4 were chosen because together they connect to all the shared disks. In practice, {project-short} first migrates VM 2 and all disks attached to it, then does the same with VM 4. Choosing VM 2 and VM 4 and only those two ensures that all the shared disks are migrated, but only once.  

Reviewer: this is messy. please can you try to clarify and break into smaller sentences, possibly use bullet points

there is also confusion; you switch from present tense `migrates` t to past tense `were chosen`, then back to present tense `migrates`.

The second `Plan` CR migrates VMs 1 and 3 and their non-shared disks. {project-short} is able to reattach all VMs to their disks, including the shared disks. 

.Example of correctly divided Plan CRs
image::OK_shared_disks.png[Example successful migration]

The second figure shows what to avoid when you divide the disks between the `Plan` CRs:

.Example of incorrectly divided Plan CRs
image::not_OK_shared_disks.png[Complex cyclic shared disk dependencies]
 

In this case, in the first `Plan` CR, {project-short} first migrates VM 2, its unshared disks, and shared disks 1 and 2. It then migrates VM 3, its unshared disks, and shared disks 2 and 3. In the second `Plan` CR, {project-short} migrates the other VMs and their unshared disks. 

> This migration "succeeds", but it results in a problem: Shared disk 2 is migrated twice by the first `Plan` CR. This problem can be resolved by using one of the two workarounds that are discussed in the Known issues section which follows the procedure.

Reviewer: ```suggestion
This migration "succeeds", but it results in a problem: Shared disk 2 is migrated twice by the first `Plan` CR. You can resolve this issue by using one of the two workarounds that are discussed in the Known issues section, which follows the procedure.
```

In this case, the VMs and shared disks cannot be migrated in the same `Plan` CR. Although this problem could be solved using `migrateSharedDisks` and 2 `Plan` CRs, it illustrates the basic issue that must be avoided in migrating VMs with shared disks.

[id="cyclic-shared-disk-workarounds_{context}"]
=== Workarounds

As discussed previously, it is important to try to create 2 `Plan` CRs in which each shared disk is migrated once. However, if your migration does result in a shared disk either duplicated or not being transferred, you can use one of the following workarounds:

* Duplicate one of the shared disks
* "Remove" one of the shared disks

[id="duplicate-shared-disk_{context}"]
==== Duplicate a shared disk

In the figure that follows, VMs 2 and 3 are migrated with the shared disks in the first plan, and VM 1 is migrated in the second plan. Doing this breaks the cyclic dependencies, but this workaround has a drawback: It results in shared disk 3 being duplicated. The solution is to remove the duplicated PV and migrate VM 1 again.

.Duplicated shared disk
image::cyclic_workaround1.png[Duplicate a shared disk]

Advantage:

The source VMs are not affected.

Disadvantage:

> One shared disk gets transferred twice.

Reviewer 2: `One shared disk gets transferred twice, and you need to manually delete the duplicate disk and reconnect VM 3 to shared disk 3 in Red Hat OpenShift after the migration.

> // Module included in the following assemblies:

Reviewer 3: General comments about the graphics.  

- The figure titles for 11.1, 11.2, and 11.3 are really long. Can we make them shorter by perhaps including more descriptive info in the content before we get to the figure?
- Did we run any of the new figures by the folks reviewing our graphics?  Every Monday you can bring new graphics to a meeting where you can get advice.  Also, we can ask others who work more with graphics for their review.
- The colors for 11.1, 11.2, and 11.3 look a bit off, and some of the arrows don't quite make their mark.

Author: @nunzy1  you're right -- see my comment above (that is, I used the wrong files)

//
// * documentation/doc-Migration_Toolkit_for_Virtualization/master.adoc

:_mod-docs-content-type: PROCEDURE
[id="mtv-shared-disks_{context}"]
= Migrating virtual machines with shared disks

You can migrate VMware virtual machines (VMs) with shared disks using the {project-first}. This functionality is available for cold migrations only. 

> Shared disks are disks which are attached to multiple VMs and which use the multi-writer option. As a result of these characteristics, shared disks are difficult to migrate. 

Reviewer 3: I use "that" in two places rather than "which".

> However, in certain situations, applications in VMs require shared disks. Databases and clustered file-systems are the primary use cases for shared disks.

Reviewer 3: I'd delete "however".  Just not needed.

> {project-short} version 2.7.11 or later includes a parameter named `migrateSharedDisks` in `Plan` CRs that instructs {project-short} to either migrate shared disks or to skip them during migration as follows:

Reviewer 3: Please follow our standards, but do we need "as follows" or does the colon suffice?

Author: @nunzy1 IBM uses "as follows:" in examples, but does not explicitly mandate including "as follows" or distinguish when to include it or not. It sounds more natural to me to include it than not, at least here, so I'm going to leave it .

> * If set to `true`, {project-short} uses the regular cold migration flow using `virt-v2v`` and labeling the shared PVC. This is the default setting. 

Reviewer 3: Two apostrophes after v2v.

> * If set to 'false`, {project-short} uses the KubeVirt CDI for disk transfer, and skips the shared disks. After the disk transfer, {project-short} automatically attempts to locate the already shared PVCs and the already migrated shared PVCs and attach them to the VMs. 

Reviewer 3: 'false`    

I might change the end of the sentence:
...and the already migrated shared PVCs, and it attaches them to the VMs.

({project-short} automatically attempts and it attaches them...)

By default, `migrateSharedDisks` is set to `true`.

> To successfully migrate VMs with shared disks, you create two `Plan` CRs as follows: 

Reviewer 3: Can leave out "you".  We might have talked about this; leave it in if it's consistent with the rest of the content. :-)

> * In the first, you set `migrateSharedDisks` to `true` and migrate the following:

Reviewer 3: Ditto, I'd delete "you" unless we need to include it.

** All shared disks

> ** For each shared disk, one of the VMs that is attached to it. If possible, choose VMs so that that plan does not contain any shared disks that are connected to more than one VM in the plan. See the figures below for further guidance.

Reviewer 3: Any of the figures below will do?  Do I need to review them all to have a good idea?  Maybe refer to one if one is more helpful than the others.

Author: @nunzy1:
 - I used the wrong files, so the figures are definitely not in line with CCS guidelines. I removed the figure for now, leaving placeholders.
 - I reduced the number of figures -- now there is one "do this" figure and one "don't do this" .

** All unshared disks attached to the VMs you choose for this plan. 

* In the second, you set `migrateSharedDisks` to 'false` and migrate the following:

** All other VMs
** The unshared disks of the VMs in the second `Plan` CR

To illustrate the division of VMs and disks between `Plan` CRs, consider the three figures that follow: 

In the first figure, the first `Plan` CR migrates VMs 1 and 3, shared disks 1, 2, and 3, and the non-shared disks of VMs 1 and 3. VMs 1 and 3 were chosen because together they connect to all the shared disks. In practice, {project-short} first migrates VM 1 and all disks attached to it, then does the same with VM 3.Choosing VM 1 and VM 3 and only those two ensures that all the shared disks are migrated, but only once.  

The second `Plan` CR migrates VMs 2 and 4 and their non-shared disks. {project-short} is able to reattach all VMs to their disks, including the shared disks. 

> .Example of a correct divsion of shared disks, VMs, and unshared disks between the 2 Plan CRs

Reviewer 3: typo -- division

image::dummy_OK1.png[Example successful migration]

In the second figure, VMs 2 and 3 and the disks they are connected to are migrated by the first `Plan` CR and the others are migrated by the second. Again, each shared VM is migrated once, despite being connected to two VMs. 

> .Additonnal example of a correct divsion of shared disks, VMs, and unshared disks between the 2 Plan CRs

Reviewer 3: typos -- additional and division

image::dummy_OK2.png[Additional example successful migration]

The third figure shows what to avoid when you divide the disks between the `Plan` CRs:

> .Example of an incorrect divsion of shared disks, VMs, and unshared disks between the 2 Plan CRs

Reviewer 3: typo - division

image::dummy_not_ok.png[Complex cyclic shared disk dependencies]

In this case, in the first `Plan` CR, {project-short} first migrates VM 1, its unshared disks, and shared disks 1 and 2. It then migrates VM 2, its unshared disks, and shared disks 2 and 3. In the second `Plan` CR, {project-short} migrates the other VMs and their unshared disks. 

> This migration "succeeds", but it results in a problem: Shared disk 2 is migrated twice by the first `Plan` CR. There are two workarounds for this problem, which are discussed in the Known issues section that follows the procedure.

Reviewer 3: I might end the last sentence after "problem", and start a new sentence with "These are discussed..."  Not an issue -- just my preference.

     uid: a3c83db3-1cf7-446a-b996-84c618946362
   storage:
     apiVersion: forklift.konveyor.io/v1beta1
     kind: StorageMap
     name: vsphere-mqp7b
     namespace: openshift-mtv
     uid: 20b43d4f-ded4-4798-b836-7c0330d552a0
 migrateSharedDisks: false
 provider:
   destination:
     apiVersion: forklift.konveyor.io/v1beta1
     kind: Provider
     name: host
     namespace: openshift-mtv
     uid: abf4509f-1d5f-4ff6-b1f2-18206136922a
   source:
     apiVersion: forklift.konveyor.io/v1beta1
     kind: Provider
     name: vsphere
     namespace: openshift-mtv
     uid: be4dc7ab-fedd-460a-acae-a850f6b9543f
 targetNamespace: openshift-mtv
 vms:
   - id: vm-71
     name: vm-2-with-shared-disks
----

. Start the migration of the second plan and wait for it to finish.
. Verify that all shared disks are attached to the same VMs as they were before migration and that none are duplicated. In case of problems, see the discussion of known issues that follows. 

[id="known-issues-shared-disks_{context}"]
== Known issues

[id="cyclic-shared-disk_{context}"]
=== Cyclic shared disk dependencies

Problem: VMs with cyclic shared disk dependencies cannot be migrated successfully.

> Explanation:  The parameter `migrateSharedDisks` only determines in terms of a migration plan if {project-short} is supposed to migrate or skip a disk. It might not recognize that a shared disk was duplicated.

Reviewer 3: When I read "only determines in terms of a migration plan", I 'm not sure what this means.  What does it determine? Can we clarify?

Also, using "might" -- As a customer, I (might) think that the parameter can't be trusted. Can we say that it does not recognize?

However, if there is a cyclic dependency of shared disks between 3 or more VMs, {project-short} either duplicates or omits one of the shared disks. The figure that follows illustrates the simplest version of this problem.

.Simple example of cyclic shared disks
image::simple-cyclic-shared-disks.png[Simple cyclic shared disk dependencies]

In this case, the VMs and shared disks cannot be migrated in the same `Plan` CR. Although this problem could be solved using `migrateSharedDisks` and 2 `Plan` CRs, it illustrates the basic issue that must be avoided in migrating VMs with shared disks.

[id="cyclic-shared-disk-workarounds_{context}"]
=== Workarounds

As discussed previously, it is important to try to create 2 `Plan` CRs in which each shared disk is migrated once. However, if your migration does result in a shared disk either duplicated or not being transferred, you can use one of the following workarounds:

* Duplicate one of the shared disks
* "Remove" one of the shared disks

[id="duplicate-shared-disk_{context}"]
==== Duplicate a shared disk

> In the figure that follows, VMs 2 and 3 are migrated with the shared disks in the first plan, and VM 1 is migrated in the second plan. Doing this breaks the cyclic dependencies, however, this workaround has a drawback: It results in shared disk 3 being duplicated. The solution to this drawback is to remove the duplicated PV and migrate VM 1 again.

Reviewer 3: "dependencies, but this workaround..."

I'd delete "to this drawback".

.Duplicated shared disk
image::cyclic_workaround1.png[Duplicate a shared disk]

Advantage:

The source VMs are not affected.

Disadvantage:

One shared disk gets transferred twice.

[id="remove-shared-link_{context}"]
==== "Remove" a shared disk

The figure that follows shows a different solution: Remove the link to one of the shared disks from one source VM. Doing this breaks the cyclic dependencies. Note that in the current VMware UI, removing the link is referred to as "removing" the disk. 

."Removed" shared disk
image::draft_workaround2.png["Remove" a shared disk]

In this case, VM 2 and 3 are migrated with the shared disks in the first plan, but the link between VM 3 and shared disk 3 is removed. As before, VM 1 is migrated in the second plan. 

> Doing this breaks the cyclic dependencies, however this workaround has its own drawback: VM 3 is disconnected from shared disk 3 and remains disconnected after the migration. The solution to this drawback is to manually reattach shared disk 3 to VM 3 after the migration finishes.

Reviewer 3: I would change however to but.

I'd delete "to this drawback".

When {project-short} migrates a VM that has a shared disk to it, it does not check if it has already migrated that shared disk. Therefore, it is important to allocate the VMs in each of the two so that each shared disk is migrated once and only once. 

To understand how to assign VMs and shared disks to each of the `Plan` CRs, consider the two figures that follow. In both, `migrateSharedDisks` is set to `true` for `plan1` and set to `false` for `plan2`.

In the first figure, the VMs and shared disks are assigned correctly:

.Example of correctly assigned VMs and shared disks
image::OK_shared_disks.png[Example successful migration]

`plan1` migrates VMs 2 and 4, shared disks 1, 2, and 3, and the non-shared disks of VMs 2 and 4. VMs 2 and 4 are included in this plan, because they connect to all the shared disks once each. 

`plan2` migrates VMs 1 and 3 and their non-shared disks. `plan2` does not migrate the shared disks connected to VMs 1 and 3 because `migrateSharedDisks` is set to `false`.

{project-short} migrates each VMs and its disks as follows:

. From `plan1`:

.. VM 3, shared disks 1 and 2, and the non-shared disks attached to VM 2.

> .. VM 4, shared disk 3, and the non-shared disks attached to VM 4. 

Reviewer: Much nicer - cleaner, easier to read, more understandable

To understand how to assign VMs and shared disks to each of the `Plan` CRs, consider the two figures that follow. In both, `migrateSharedDisks` is set to `true` for `plan1` and set to `false` for `plan2`.

In the first figure, the VMs and shared disks are assigned correctly:

.Example of correctly assigned VMs and shared disks
image::OK_shared_disks.png[Example successful migration]

`plan1` migrates VMs 2 and 4, shared disks 1, 2, and 3, and the non-shared disks of VMs 2 and 4. VMs 2 and 4 are included in this plan, because they connect to all the shared disks once each. 

`plan2` migrates VMs 1 and 3 and their non-shared disks. `plan2` does not migrate the shared disks connected to VMs 1 and 3 because `migrateSharedDisks` is set to `false`.

{project-short} migrates each VMs and its disks as follows:

. From `plan1`:

.. VM 3, shared disks 1 and 2, and the non-shared disks attached to VM 2.
.. VM 4, shared disk 3, and the non-shared disks attached to VM 4. 

. From `plan2`:

.. VM 1 and the non-shared disks attached to it.

> .. VM 2 and the non-shared disks attached to it.

Reviewer: :+1:
